### ISSUE TITLE
XmlWriter produces formally incorrect XML

### ISSUE THREAD

**Issue by Lercher** [clbanning/mxj#88]
But it should not, of course. E.g. serializing this mxj-map

```go
        m[""] = "empty"
	m["$invalid"] = "hex$"
	m["entities"] = "<>&"
	m["nil"] = nil
```

produces this string, which is *formally* incorrect XML:

```txt
<root>
  <>empty</>
  <$invalid>hex$</$invalid>
  <entities><>&</entities>
  <nil><nil></nil>
</root>
```

Errors are:

* `<>` is not a valid XML element
* `$` is not valid in XML element names
* the three basic XML entities `<`, `>` and `&` must be escaped as `&lt;`, `&gt;` and `&amp;`
* the `nil` `interface{}` value is serialized as `<nil>` without escaping `<` and `>`

Additionally there are some decisions to be drawn for how to mitigate these problems. Probably some strategies need to be implemented for a client to customize the behavior of the API

* empty element names should either be omitted or given a customizeable name
* invalid element name characters could be either dropped completely or somehow encoded by valid entity characters like a hex string representation decorated with the valid `_` char
* nil valued elements could be serialized as `<element-name/>`, `<element-name></element-name>`, omitted, or marked with `xsi:nil="true"` (see e.g. https://www.oreilly.com/library/view/xml-in-a/0596007647/re166.html)

## test fixture (example)

```go
package web

import (
	"bytes"
	"testing"

	"github.com/clbanning/mxj/v2"
)

func TestJSON2XMLConverter(t *testing.T) {
	m := mxj.New()

	m[""] = "empty"
	m["$invalid"] = "hex$"
	m["entities"] = "<>&"
	m["nil"] = nil

	buf := new(bytes.Buffer)
	err := m.XmlWriter(buf, "root")
	if err != nil {
		t.Fatalf("writer: %v", err)
	}

	if got, want := buf.String(), `<root><_24_invalid>hex$</_24_invalid><entities>&lt;&gt;&amp;</entities><nil></nil></root>`; got != want {
		// want <root><_>empty</_><_24_invalid>hex$</_24_invalid><entities>&lt;&gt;&amp;</entities><nil></nil></root>
		// got <root><>empty</><$invalid>hex$</$invalid><entities><>&</entities><nil><nil></nil></root>
		t.Errorf("want %v, got %v", want, got)
	}
}
```

**Comment by Lercher** [comment]
Sorry, I've overlooked this global variable `mxj.XMLEscapeChars(true)` which tackles the entities problem. It is very problematic in multi-threaded systems like a common http server, however.

**Comment by clbanning** [comment]
Looks like an issue possibly appropriate for clbanning/anyxml package; here I'm not so sure if it involves interrogating the Map values.  Though ```m[""]<value>``` and ```m["nil"]nil``` could be flagged.  

Please give me a use-case where a production program would create any of these values, particularly via ```mxj.NewMapXml``` or ```mxj.NewMapJson```, which are the common functions used to create ```mxj.Map``` values.

**Comment by Lercher** [comment]
Productive use case, of course: penetration test

Create the initial map as a JS object on the browser, `JSON.stringify` it, post it to a go based GUI server, deserializes it with the `json` package to a generic map, convert it with `mxj` to „xml“ and send that string to an xml processing backend server, which hopefully errors out with „invalid XML“.

**Comment by Lercher** [comment]
Well, yes `if _, err := mxj.NewMapXml(buf.Bytes()); err != nil {` in the last block errors with `err: xml.Decoder.Token() - XML syntax error on line 1: expected element name after <` because `err = m.XmlWriter(buf, "root")` didn't write formally valid XML. That's exactly this ticket's topic.

N.B.: `XML: <root><>empty</><$invalid>hex$</$invalid><entities><>&</entities><nil><nil></nil></root>`

**Comment by clbanning** [comment]
Patched at tip, v2.5.0, by toggling on XmlCheckIsValid().
